scalding-core/src/main/scala/com/twitter/scalding/ExecutionOptimizationRules.scala

```diff
@@ -419,16 +249,17 @@ object ExecutionOptimizationRules {
      * into a single value. If ex is already optimal (0 or 1 write) return None
      */
     def optimize[A](ex: Execution[A]): Option[Execution[A]] = {
-      val flat = FlattenedZip(ex)
-      val wc = flat.writeCount
-      // only optimize if there are 2 or more writes, otherwise we create an infinite loop
-      if (wc > 1) {
-        // unless there is a bug, this optimized will be defined
-        // but internally, optimize can return None. Just to be
-        // cleaner we use flatMap here rather than getOrElse(sys.error("unreachable"))
-        flat.optimize.map(_.execution)
+      def writes(ex: Execution[_]): Int = ex match {
```

Comment:
can we make this `@annotation.tailrec` so we can check this without blowing the stack?

e.g.

```scala
@annotation.tailrec
def writes(ex: List[Execution[Any]], acc: Int): Int =
  ex match {
    case Nil => acc
    case h :: tail =>
     ...
```

Reply:
I've tried it but it doesn't work because of `zip` - it sums two branches so `tailrec` isn't applicable.

Reply:
I think this works:
```scala
@annotation.tailrec
def writes(ex: List[Execution[Any]], acc: Int): Int =
  ex match {
    case Nil => acc
    case h :: tail =>
      h match {
        case Zipped(l, r) => writes(l :: r :: tail, acc)
        case Mapped(e, _) => writes(e :: tail, acc)
        case WriteExecution(_, _, _) => writes(tail, acc + 1)
        case _ => writes(tail, acc)
  }
```
then we can define `def writes(ex: Execution[Any]): Int = writes(ex :: Nil, 0)`

Won't that work?

Reply:
It works, thanks! This is interesting way of making recursion tail. Really liked it!